integration/container/stop_test.go

```diff
@@ -42,6 +43,61 @@ func TestStopContainerWithRestartPolicyAlways(t *testing.T) {
 	}
 }
 
+// TestStopContainerWithTimeout checks that ContainerStop with
+// a timeout works as documented, i.e. in case of negative timeout
+// waiting is not limited (issue #35311).
+func TestStopContainerWithTimeout(t *testing.T) {
+	defer setupTest(t)()
+	client := request.NewAPIClient(t)
+	ctx := context.Background()
+
+	testCmd := container.WithCmd("sh", "-c", "sleep 2 && exit 42")
+	testData := []struct {
+		doc              string
+		timeout          int
+		expectedExitCode int
+	}{
+		// In case container is forcefully killed, 137 is returned,
+		// otherwise the exit code from the above script
+		{
+			"zero timeout: expect forceful container kill",
+			0, 137,
+		},
+		{
+			"too small timeout: expect forceful container kill",
+			1, 137,
+		},
+		{
+			"big enough timeout: expect graceful container stop",
+			3, 42,
+		},
+		{
+			"unlimited timeout: expect graceful container stop",
+			-1, 42,
+		},
+	}
+
+	for _, d := range testData {
+		d := d
+		name := t.Name() + "_" + strconv.Itoa(d.timeout)
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
```

Comment:
I think each one of these is treated as a separate test, so they all get grouped together at the end of the test run with the rest of the parallel tests. So I'm not sure this is any safer than before.

Reply:
1. I think this is not so.  According to the [golang pkg/testing documentation](https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks), such tests are only run in parallel with each other (emphasize is mine):

<blockquote>
Subtests can also be used to control parallelism. A parent test will only complete once all of its subtests complete. In this example, all tests are run in parallel with each other, and only with each other, regardless of other top-level tests that may be defined:

```go
func TestGroupedParallel(t *testing.T) {
    for _, tc := range tests {
        tc := tc // capture range variable
        t.Run(tc.Name, func(t *testing.T) {
            t.Parallel()
            ...
        })
    }
}
```
</blockquote>

2. I got the idea from similar tests like this one: https://github.com/moby/moby/blob/fe2d3a1551f666465e65ec4337e92b774a5e81f1/integration/container/create_test.go#L41-L46